changelog/range-error.md

```diff
@@ -0,0 +1,54 @@
+Out of bounds array access now gives a fluent error message
+
+Errors resulting from bad indexing will now contain the length of
+the array, as well as the offending index for `arr[bad]`,
+or offending indices `arr[bad1 .. bad2]` for bad slices.
+
+For example:
+
+---
+void main()
+{
+    int[] a = [1, 2, 3];
+    int b = a[7];
+}
+---
+
+Previously this would yield the following error when compiled and run:
+
+$(CONSOLE
+> dmd -run main.d
+core.exception.RangeError@main.d(4): Range violation
+$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)
+??:? _d_arrayboundsp [0x555765c167f9]
+??:? _Dmain [0x555765c16752]
+)
+
+It now yields:
+
+$(CONSOLE
+> dmd -run main.d
+core.exception.ArrayIndexError@main.d(4): index [7] exceeds array of length 3
+$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)
+??:? _d_arraybounds_indexp [0x5647250b980d]
+??:? _Dmain [0x5647250b9751]
+)
+
+Similarly, in case of out of bounds slice operations:
+---
+void main()
+{
+    int[] a = [1, 2, 3];
+    int b = a[2 .. 4];
+}
+---
+
+$(CONSOLE
+> dmd -run main.d
+core.exception.ArraySliceError@main.d(4): slice [2 .. 4] extends past array of length 3
+$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)
+??:? _d_arraybounds_slicep [0x5647250b980d]
+??:? _Dmain [0x5647250b9751]
+)
+
+The error message for indexing a non-existent key in an Associative Array has not been updated.
```

Comment:
Not sure why this mention was warranted ?

Reply:
I figured people might wonder if all range errors are improved, so I'm stating a known limitation here.